### ISSUE TITLE
Add cancellation for selected filters?

### ISSUE THREAD

**Issue by elvisyjlin** [pschydlo/RLenv.directory#3]
Hi Paul, 

Thank you for making a page for RL repos. I find it useful when I want to explore more on RL. While browsing the page, I found it slightly inconvenient that the selected filters (agents and complexity) cannot be removed as I wish. 

Perhaps a modification like the following image shows would be fine:
<img width="1406" alt="screen shot 2018-09-03 at 10 21 35 pm" src="https://user-images.githubusercontent.com/10172439/44992437-c7ca2080-afc9-11e8-943a-612f6f1e2349.png">

I'm willing to help if you consider it a good feature to add in and thanks again for this awesome page.

**Comment by pschydlo** [comment]
Hey Elvis!

Thanks for the kind feedback and excellent suggestion!

I have been thinking about a filter reset button for a while now, but your solution is much more elegant!

At the moment the control panel (site/webpack/components/ControlPanel.jsx) sends the filter options to the App component (App.jsx) through the filterChange callback method from the App component. This method then adds the new options to the filter, which means its only additive. 

A solution would be to remove the property from the filter if the ControlPanel sends this entry as None.

```
 filterChange(new_filter_opt){

    var filter_opt = this.state.filter_opt 

    for (var key in new_filter_opt){
      filter_opt[key] = new_filter_opt[key]
    }

(...)
```
 
We could just do something like this:

```
if(filter_opt[key] == None){
   delete key from filter_opt
}
```

It's a tiny bit hacky, but until I refactor it to become more flexible and elegant I think this is our best solution!

Wish you a great day!
Thanks again for the great idea!

all the best, 
Paul

**Comment by pschydlo** [comment]
Just finished refactoring how the child components send filter options to the parent component. 

Now instead of sending a dictionary with the options (which the parent has to interpret), the child component sends the filter functions directly. This helps manage the complexity of the parent component (App), which already has a lot of responsibilities.

For example, complexity selection now looks like this:

```
// Callback function for agent dropdown selection

complexitySelect(level){
  var level = level.toLowerCase()

  var filterFn = function(env)
  { 
    if (env.complexity != level) return false 
    return true
  }

  this.updateFilterOptions('complexity', filterFn)
}
```
Now we could directly change the filterFn to return True if the "All" level is selected.

```
  var filterFn = function(env)
  { 
    if(level=='all') return true
    if (env.complexity != level) return false 
    return true
  }
```

**Comment by elvisyjlin** [comment]
I was wondering how to host the site on my computer. I appreciate for your clear instructions! It helps a lot. I'll do it!

**Comment by pschydlo** [comment]
Thank you! :) If you run into any kind of questions regarding the code feel free to ask! I'll try to answer as soon as possible!

All the best,
Paul

**Comment by elvisyjlin** [comment]
I've done the modifications of filter reset button ("All") for agent and complex. What I revised are `ControlPanel.jsx` and `DropBot.jsx`. Should I commit the newly generated `bundle.js` as well?

By the way, in order to host the site locally, I did a little bit more than installing webpack and jekyll. It said Ruby lacked `pygments` to run jekyll, so I added `gem "pygments.rb"`in the Gemfile and do `bundle install` again. And my jekyll said `-l` is an ambiguous option which prevent my to host the site, hence I run it with merely `jekyll serve` and everything works well finally.

**Comment by pschydlo** [comment]
Perfect, that sounds about right!

Glad to hear you were able to run the site locally!

Thank you for the hint regarding the Pygments gem, it seems to be windows specific, did it work after adding the gem to the Gemfile? Were there any additional steps you had to perform?

Regarding the `jekyll serve -l` command, the `-l` option is meant to live reload the page as you edit the assets, does the option show up when you execute `jekyll serve --help` ?

It should be safe to commit the regenerated `bundle.js`.

Did you come across any other issues or architecture change suggestions?

Looking forward to adding you as a proud contributor to our lovely about page: 
https://rlenv.directory/about/ !

Wish you an excellent week,
Paul

**Comment by elvisyjlin** [comment]
I'll open a pull request soon.

I could hardly remember the exact steps because I tried 3 times (delete the folder and clone again) to host successfully haha... but I believe the key step is the pygments stuff. I'm using MacOS btw. Another thing I did was adding `exclude: [vendor, node_modules]` to the _config.yml file. Otherwise, I get error messages when running jekyll.

I installed jekyll by the following command under the site folder
```bash
npm install jekyll
```  
After running
```bash
node_modules/jekyll/bin/jekyll serve --help
```
I got
```text
Usage:

  jekyll serve [options]

Options:
            --config CONFIG_FILE[,CONFIG_FILE2,...]  Custom configuration file
            --future       Publishes posts with a future date
            --limit_posts MAX_POSTS  Limits the number of posts to parse and publish
        -w, --[no-]watch   Watch for changes and rebuild
            --force_polling  Force watch to use polling
            --lsi          Use LSI for improved related posts
        -D, --drafts       Render posts in the _drafts folder
            --unpublished  Render posts that were marked as unpublished
        -q, --quiet        Silence output.
        -V, --verbose      Print verbose output.
        -f, --full-rebuild  Disable incremental rebuild.
        -B, --detach       Run the server in the background (detach)
        -P, --port [PORT]  Port to listen on
        -H, --host [HOST]  Host to bind to
        -b, --baseurl [URL]  Base URL
            --skip-initial-build  Skips the initial site build which occurs before the server is started.
        -h, --help         Show this message
        -v, --version      Print the name and version
        -t, --trace        Show the full backtrace when an error occurs
        -s, --source [DIR]  Source directory (defaults to ./)
        -d, --destination [DIR]  Destination directory (defaults to ./_site)
            --safe         Safe mode (defaults to false)
        -p, --plugins PLUGINS_DIR1[,PLUGINS_DIR2[,...]]  Plugins directory (defaults to ./_plugins)
            --layouts DIR  Layouts directory (defaults to ./_layouts)
        -h, --help         Show this message
        -v, --version      Print the name and version
        -t, --trace        Show the full backtrace when an error occurs
```
Weird. There is no `-l` option... Anyway, your code is very well-contructed. I'm able to understand the .jsx scripts, even though I'm new to React. Please test my modification carefully. I'll change anything that you think hacky.

Thank you so much!

**Comment by pschydlo** [comment]
I think I found out the Jekyll problem, it looks like there are two ways of installing it, either as a Gem or through NPM, I have it installed as a Gem which would explain the difference in behavior. 

I'll add a note regarding the Jekyll version in the [contributor guide I'm writing](https://github.com/pschydlo/RLenv.directory/blob/master/CONTRIBUTING.md)! Thanks again for the hint!

For future reference regarding the issue: This feature request was addressed successfully in the pull request #4
